local/src/core/http2.cc

```diff
@@ -698,10 +698,48 @@ receive_nghttp2_request(
 
 static int
 on_frame_send_cb(
-    nghttp2_session * /* session */,
-    nghttp2_frame const * /* frame */,
-    void * /* user_data */)
+    nghttp2_session *session,
+    nghttp2_frame const *frame,
+    void * user_data)
 {
+  auto *session_data = reinterpret_cast<H2Session *>(user_data);
+  auto const stream_id = frame->hd.stream_id;
+  auto stream_map_iter = session_data->_stream_map.find(stream_id);
+  if (stream_map_iter == session_data->_stream_map.end()) {
+    // Nothing to do if this is not in our stream map.
+    return 0;
+  }
+  H2StreamState &stream_state = *stream_map_iter->second;
+  Errata errata;
+
+  errata.note(S_DIAG, "Sent frame: {}", H2FrameNames[static_cast<H2Frame>(frame->hd.type)]);
+  if (stream_id != 0) {
+    if (frame->hd.type == stream_state._client_abort_frame) {
+      errata.note(
+        S_DIAG,
+        "Testing client abort, submitting RST_STREAM frame for key {} with error code {}.",
+        stream_state._key,
+        H2ErrorCodeNames[static_cast<H2ErrorCode>(stream_state._client_abort_error)]
+      );
+      errata.sink();
+      int const submit_status = nghttp2_submit_rst_stream(session, 0, stream_id, stream_state._client_abort_error);
+      if (submit_status) {
+        errata.note(S_DIAG, "Failed to submit RST_STREAM frame: {}.", submit_status);
+      }
+    } else if (frame->hd.type == stream_state._server_abort_frame) {
+      errata.note(
+        S_DIAG,
+        "Testing server abort, submitting RST_STREAM frame for key {} with error code {}.",
+        stream_state._key,
+        H2ErrorCodeNames[static_cast<H2ErrorCode>(stream_state._server_abort_error)]
+      );
+      errata.sink();
+      int const submit_status = nghttp2_submit_rst_stream(session, 0, stream_id, stream_state._server_abort_error);
+      if (submit_status) {
+        errata.note(S_DIAG, "Failed to submit RST_STREAM frame: {}.", submit_status);
```

Comment:
Good to provide the key here. I've found from painful experience that any logs like this must have the key to be helpful. In the traffic reply case, there can be thousands of these and it's hard to distinguish them otherwise.